gprofiler/spark/spark_collector.py

```diff
@@ -0,0 +1,899 @@
+        applications_metrics: bool = True,
+        streaming_metrics: bool = True,
+    ) -> None:
+        self._lock = Lock()
+        self._logger = get_logger_adapter(__name__)
+
+        self._last_sample_time = int(time.monotonic() * 1000)
+        self._cluster_mode = cluster_mode
+        self._master_address = f"http://{master_address}"
+        self._cluster_metrics = cluster_metrics
+        self._applications_metrics = applications_metrics
+        self._streaming_metrics = streaming_metrics
+        self._task_summary_metrics = True
+        self._last_iteration_app_job_metrics: Dict[str, Dict[str, Any]] = {}
+
+    def collect(self) -> Generator[Dict[str, Any], None, None]:
+        if not self._lock.acquire(False):
+            self._logger.warning(
+                "Could not acquire collector's mutex, increasing scrape interval should solve this",
+                extra={"collector": self.__class__.__name__},
+            )
+            return
+        try:
+            for data in self._collect():
+                yield data
+        except Exception:
+            self._logger.exception(f"Error while trying to collect f{self.__class__.__name__}")
+        finally:
+            self._lock.release()
+
+    def _collect(self) -> Generator[Dict[str, Any], None, None]:
+        try:
+            collected_metrics: Dict[str, Dict[str, Any]] = {}
+
+            if self._cluster_metrics:
+                if self._cluster_mode == "yarn":
+                    self._yarn_cluster_metrics(collected_metrics)
+                    self._yarn_nodes_metrics(collected_metrics)
+                elif self._cluster_mode == "driver":
+                    pass
+
+            if self._applications_metrics:
+                spark_apps = self._get_running_apps()
+                self._spark_application_metrics(collected_metrics, spark_apps)
+                self._spark_stage_metrics(collected_metrics, spark_apps)
+                self._spark_executor_metrics(collected_metrics, spark_apps)
+                if self._streaming_metrics:
+                    self._spark_batches_streams_metrics(collected_metrics, spark_apps)
+                    self._spark_streaming_statistics_metrics(collected_metrics, spark_apps)
+                    self._spark_structured_streams_metrics(collected_metrics, spark_apps)
+
+                # Get the rdd metrics
+                self._spark_rdd_metrics(collected_metrics, spark_apps)
+
+            for metric in collected_metrics.values():
+                yield metric
+
+            self._logger.debug("Succeeded gathering spark metrics")
+
+        except Exception:
+            self._logger.warning("Error while trying collect spark metrics")
+            self._logger.debug(
+                traceback.format_exc()
+            )  # spark collector exceptions tend to be very verbose, so print only in debug
+
+        finally:
+            self._last_sample_time = int(time.time() * 1000)  # need to be in ms
```

Comment:
time.monotonic same as in the c'tor. Or just add a function `update_sample_time()`